src/ray/raylet/worker_pool.cc

```diff
@@ -591,6 +593,64 @@ void WorkerPool::PushWorker(const std::shared_ptr<WorkerInterface> &worker) {
   }
 }
 
+void WorkerPool::TryToKillWorker(std::shared_ptr<WorkerInterface> worker) {
+  auto &state = GetStateForLanguage(worker->GetLanguage());
+  if (state.killed_workers.count(worker) > 0) {
+    // This worker has already been killed.
+    // This is possible because a Java worker process may hold multiple workers.
+    return;
+  }
+
+  auto registered_size = GetAllRegisteredWorkers().size();
+  for (const auto &entry : states_by_lang_) {
+    registered_size -= entry.second.killed_workers.size();
+  }
+  if (registered_size <= static_cast<size_t>(num_workers_soft_limit_)) {
+    return;
+  }
+
+  auto worker_id = worker->WorkerId();
+  const auto pid = worker->GetProcess().GetId();
+  if (state.idle.count(worker) == 0) {
+    return;
+  }
+  if (state.starting_worker_processes.count(worker->GetProcess()) > 0) {
+    // A Java worker process may hold multiple workers.
+    RAY_LOG(DEBUG) << "Some workers of pid " << pid
+                   << " are pending registration. Skip killing worker " << worker_id;
+    return;
+  }
+
+  // Make sure all workers in this worker process are idle.
+  // This block of code is needed by Java workers.
+  std::unordered_set<std::shared_ptr<WorkerInterface>> workers_in_the_same_process;
+  for (const auto &worker_in_the_same_process : state.registered_workers) {
+    if (worker_in_the_same_process->GetProcess().GetId() == pid) {
+      if (state.idle.count(worker_in_the_same_process) == 0) {
+        // Another worker in this process isn't idle, so this process can't be killed.
+        return;
+      } else {
+        workers_in_the_same_process.insert(worker_in_the_same_process);
+      }
+    }
+  }
+
+  for (auto worker_it = workers_in_the_same_process.begin();
+       worker_it != workers_in_the_same_process.end(); worker_it++) {
+    RAY_LOG(INFO) << "The worker pool has " << registered_size
+                  << " registered workers which exceeds the soft limit of "
+                  << num_workers_soft_limit_ << ", and worker "
+                  << (*worker_it)->WorkerId() << " with pid " << pid
+                  << " is idle. Kill it.";
+    // Remove the worker from the idle pool so it can't be popped anymore. However, we
+    // don't remove it from the registered pool because we want the worker to go through
+    // the normal disconnection logic in Node Manager.
+    RemoveWorker(state.idle, *worker_it);
+    state.killed_workers.insert(*worker_it);
```

Comment:
Why do we need killed_workers? Is it because without this we are losing a shared pointer of a worker? 

If so, I feel like the name killed_workers is kind of misleading because we are killing workers outside this method. For example, placement group removal kills workers of that placement group.

Maybe we should a more clear term like `killed_idle_workers`?

Reply:
It's used for counting only. Because the worker is removed from registered workers only when the connection breaks, the condition to trigger capping is "size(registered workers) - size(killed workers) > soft limit". I agree the name is not straightforward enough. Maybe rename it to 'pending_unregistration_workers' refine the comment on this field?

Reply:
SGTM!